File: src/main/java/com/weeth/domain/board/domain/repository/PostRepository.java
package com.weeth.domain.board.domain.repository;

import com.weeth.domain.board.domain.entity.Post;
import com.weeth.domain.board.domain.entity.enums.Category;
import com.weeth.domain.board.domain.entity.enums.Part;
import jakarta.persistence.LockModeType;
import jakarta.persistence.QueryHint;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Slice;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Lock;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.jpa.repository.QueryHints;
import org.springframework.data.repository.query.Param;

import java.util.Collection;
import java.util.List;

public interface PostRepository extends JpaRepository<Post, Long> {

	@Lock(LockModeType.PESSIMISTIC_WRITE)
	@QueryHints(@QueryHint(name = "jakarta.persistence.lock.timeout", value = "2000"))
	@Query("select p from Post p where p.id = :id")
	Post findByIdWithLock(@Param("id") Long id);

	@Query("""
        SELECT p FROM Post p
        WHERE p.category IN (
            com.weeth.domain.board.domain.entity.enums.Category.StudyLog,
            com.weeth.domain.board.domain.entity.enums.Category.Article
        )
        ORDER BY p.id DESC
    """)
	Slice<Post> findRecentPart(Pageable pageable);

	@Query("""
        SELECT p FROM Post p
        WHERE p.category = com.weeth.domain.board.domain.entity.enums.Category.Education
        ORDER BY p.id DESC
    """)
	Slice<Post> findRecentEducation(Pageable pageable);

	@Query("""
        SELECT p FROM Post p
        WHERE p.category IN (
            com.weeth.domain.board.domain.entity.enums.Category.StudyLog,
            com.weeth.domain.board.domain.entity.enums.Category.Article
        )
          AND (
                LOWER(p.title)   LIKE LOWER(CONCAT('%', :kw, '%'))
             OR LOWER(p.content) LIKE LOWER(CONCAT('%', :kw, '%'))
          )
        ORDER BY p.id DESC
    """)
	Slice<Post> searchPart(@Param("kw") String kw, Pageable pageable);

	@Query("""
        SELECT p FROM Post p
        WHERE p.category = com.weeth.domain.board.domain.entity.enums.Category.Education
          AND (
                LOWER(p.title)   LIKE LOWER(CONCAT('%', :kw, '%'))
             OR LOWER(p.content) LIKE LOWER(CONCAT('%', :kw, '%'))
          )
        ORDER BY p.id DESC
    """)
	Slice<Post> searchEducation(@Param("kw") String kw, Pageable pageable);

	@Query("""
		SELECT DISTINCT p.studyName
		FROM Post p
		WHERE (:part = com.weeth.domain.board.domain.entity.enums.Part.ALL OR p.part = :part)
		  AND p.studyName IS NOT NULL
		ORDER BY p.studyName ASC
	""")
	List<String> findDistinctStudyNamesByPart(@Param("part") Part part);

	@Query("""
        SELECT p
          FROM Post p
         WHERE (p.part = :part OR p.part = com.weeth.domain.board.domain.entity.enums.Part.ALL OR :part = com.weeth.domain.board.domain.entity.enums.Part.ALL
         )
           AND (:category IS NULL OR p.category = :category)
           AND (:cardinal IS NULL OR p.cardinalNumber = :cardinal)
           AND (:studyName IS NULL OR p.studyName = :studyName)
           AND (:week IS NULL OR p.week = :week)
      ORDER BY p.id DESC
    """)
	Slice<Post> findByPartAndOptionalFilters(@Param("part") Part part, @Param("category") Category category, @Param("cardinal") Integer cardinal, @Param("studyName") String studyName, @Param("week") Integer week, Pageable pageable);

	@Query("""
		SELECT p
		  FROM Post p
		 WHERE p.category = :category
		   AND (:cardinal IS NULL OR p.cardinalNumber = :cardinal)
		   AND (
				 :partName = 'ALL'
			  OR FUNCTION('FIND_IN_SET', :partName, p.parts) > 0
			  OR FUNCTION('FIND_IN_SET', 'ALL',    p.parts) > 0
			   )
	  ORDER BY p.id DESC
	""")
	Slice<Post> findByCategoryAndOptionalCardinalWithPart(@Param("partName") String partName, @Param("category") Category category, @Param("cardinal") Integer cardinal, Pageable pageable);

	@Query("""
		SELECT p
		  FROM Post p
		 WHERE p.category = :category
		   AND p.cardinalNumber = :cardinal
		   AND (
				 :partName = 'ALL'
			  OR FUNCTION('FIND_IN_SET', :partName, p.parts) > 0
			  OR FUNCTION('FIND_IN_SET', 'ALL',    p.parts) > 0
			 )
	  ORDER BY p.id DESC
	""")
	Slice<Post> findByCategoryAndCardinalNumberWithPart(@Param("partName") String partName, @Param("category") Category category, @Param("cardinal") Integer cardinal, Pageable pageable);

	@Query("""
		SELECT p
		  FROM Post p
		 WHERE p.category = :category
		   AND p.cardinalNumber IN :cardinals
		   AND (
				 :partName = 'ALL'
			  OR FUNCTION('FIND_IN_SET', :partName, p.parts) > 0
			  OR FUNCTION('FIND_IN_SET', 'ALL',    p.parts) > 0
			 )
	  ORDER BY p.id DESC
	""")
	Slice<Post> findByCategoryAndCardinalInWithPart(@Param("partName") String partName, @Param("category") Category category, @Param("cardinals") Collection<Integer> cardinals, Pageable pageable);
}

File: src/main/kotlin/com/weeth/domain/board/domain/repository/PostRepository.kt
package com.weeth.domain.board.domain.repository

import com.weeth.domain.board.application.exception.PostNotFoundException
import com.weeth.domain.board.domain.entity.Post
import com.weeth.domain.board.domain.enums.BoardType
import jakarta.persistence.LockModeType
import jakarta.persistence.QueryHint
import org.springframework.data.domain.Page
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Pageable
import org.springframework.data.domain.Slice
import org.springframework.data.jpa.repository.EntityGraph
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Lock
import org.springframework.data.jpa.repository.Query
import org.springframework.data.jpa.repository.QueryHints
import org.springframework.data.repository.query.Param
import java.time.LocalDateTime

interface PostRepository :
    JpaRepository<Post, Long>,
    PostReader {
    @EntityGraph(attributePaths = ["clubMember", "clubMember.user", "clubMember.userProfile", "board"])
    @Query(
        """
        SELECT p
        FROM Post p
        WHERE p.board.id IN :boardIds
          AND p.isDeleted = false
          AND p.board.isDeleted = false
        ORDER BY p.createdAt DESC, p.id DESC
        """,
    )
    fun findAllActiveByBoardIds(
        @Param("boardIds") boardIds: List<Long>,
        pageable: Pageable,
    ): Slice<Post>

    @EntityGraph(attributePaths = ["clubMember", "clubMember.user", "clubMember.userProfile", "board"])
    @Query(
        """
        SELECT p
        FROM Post p
        WHERE p.board.id = :boardId
          AND p.isDeleted = false
          AND p.board.isDeleted = false
        """,
    )
    fun findAllActiveByBoardId(
        @Param("boardId") boardId: Long,
        pageable: Pageable,
    ): Slice<Post>

    fun findByIdAndIsDeletedFalse(id: Long): Post?

    @EntityGraph(attributePaths = ["clubMember", "clubMember.user", "clubMember.userProfile", "board"])
    @Query(
        """
        SELECT p
        FROM Post p
        WHERE p.id = :id
          AND p.isDeleted = false
          AND p.board.isDeleted = false
        """,
    )
    fun findActivePostById(
        @Param("id") id: Long,
    ): Post?

    @EntityGraph(attributePaths = ["board", "board.club"])
    @Lock(LockModeType.PESSIMISTIC_WRITE)
    @QueryHints(QueryHint(name = "jakarta.persistence.lock.timeout", value = "2000"))
    @Query(
        """
        SELECT p
        FROM Post p
        WHERE p.id = :id
          AND p.isDeleted = false
          AND p.board.isDeleted = false
        """,
    )
    fun findByIdWithLock(
        @Param("id") id: Long,
    ): Post?

    @EntityGraph(attributePaths = ["board", "board.club"])
    @Lock(LockModeType.PESSIMISTIC_WRITE)
    @QueryHints(QueryHint(name = "jakarta.persistence.lock.timeout", value = "2000"))
    @Query(
        """
        SELECT p
        FROM Post p
        WHERE p.id IN :ids
          AND p.isDeleted = false
          AND p.board.isDeleted = false
        ORDER BY p.id ASC
        """,
    )
    fun findAllByIdsWithLock(
        @Param("ids") ids: List<Long>,
    ): List<Post>

    @EntityGraph(attributePaths = ["clubMember", "clubMember.user", "clubMember.userProfile", "board"])
    @Query(
        """
        SELECT p
        FROM Post p
        WHERE p.board.id = :boardId
          AND p.isDeleted = false
          AND p.board.isDeleted = false
          AND (LOWER(p.title) LIKE LOWER(CONCAT('%', :keyword, '%')) OR LOWER(p.content) LIKE LOWER(CONCAT('%', :keyword, '%')))
        """,
    )
    fun searchByBoardId(
        @Param("boardId") boardId: Long,
        @Param("keyword") keyword: String,
        pageable: Pageable,
    ): Slice<Post>

    override fun getById(postId: Long): Post = findActivePostById(postId) ?: throw PostNotFoundException()

    override fun findActiveById(postId: Long): Post? = findActivePostById(postId)

    override fun findRecentByBoardIds(
        boardIds: List<Long>,
        pageable: Pageable,
    ): Slice<Post> = findAllActiveByBoardIds(boardIds, pageable)

    @EntityGraph(attributePaths = ["clubMember", "clubMember.user", "clubMember.userProfile"])
    @Query(
        """
        SELECT p
        FROM Post p
        WHERE p.board.type = :boardType
          AND p.isDeleted = false
          AND p.board.isDeleted = false
        ORDER BY p.createdAt DESC, p.id DESC
        """,
    )
    override fun findRecentByBoardType(
        @Param("boardType") boardType: BoardType,
        pageable: Pageable,
    ): Slice<Post>

    @EntityGraph(attributePaths = ["clubMember", "clubMember.user", "clubMember.userProfile"])
    @Query(
        """
        SELECT p
        FROM Post p
        WHERE p.board.type <> :excludedType
          AND p.isDeleted = false
          AND p.board.isDeleted = false
        ORDER BY p.createdAt DESC, p.id DESC
        """,
    )
    override fun findRecentExcludingBoardType(
        @Param("excludedType") excludedType: BoardType,
        pageable: Pageable,
    ): Slice<Post>

    @EntityGraph(attributePaths = ["clubMember", "clubMember.user", "clubMember.userProfile"])
    @Query(
        """
        SELECT p
        FROM Post p
        WHERE p.board.club.id = :clubId
          AND p.board.type = :boardType
          AND p.isDeleted = false
          AND p.board.isDeleted = false
        ORDER BY p.createdAt DESC, p.id DESC
        """,
    )
    override fun findRecentByClubIdAndBoardType(
        @Param("clubId") clubId: Long,
        @Param("boardType") boardType: BoardType,
        pageable: Pageable,
    ): Slice<Post>

    @EntityGraph(attributePaths = ["clubMember", "clubMember.user", "clubMember.userProfile"])
    @Query(
        """
        SELECT p
        FROM Post p
        LEFT JOIN LastNoticeRead lr ON lr.clubMember.id = :clubMemberId AND lr.board.id = p.board.id
        WHERE p.board.club.id = :clubId
          AND p.board.type = :boardType
          AND p.isDeleted = false
          AND p.board.isDeleted = false
          AND p.createdAt >= :since
          AND (lr IS NULL OR p.createdAt > lr.lastReadAt)
        ORDER BY p.createdAt DESC, p.id DESC
        """,
    )
    fun findUnreadNoticeSince(
        @Param("clubId") clubId: Long,
        @Param("clubMemberId") clubMemberId: Long,
        @Param("boardType") boardType: BoardType,
        @Param("since") since: LocalDateTime,
        pageable: Pageable,
    ): List<Post>

    override fun findFirstUnreadNoticeSince(
        clubId: Long,
        clubMemberId: Long,
        boardType: BoardType,
        since: LocalDateTime,
    ): Post? = findUnreadNoticeSince(clubId, clubMemberId, boardType, since, PageRequest.of(0, 1)).firstOrNull()

    @Query(
        """
        SELECT new com.weeth.domain.board.domain.repository.BoardPostCount(p.board.id, COUNT(p))
        FROM Post p
        WHERE p.board.id IN :boardIds
          AND p.isDeleted = false
        GROUP BY p.board.id
        """,
    )
    fun countActivePostsByBoardIds(
        @Param("boardIds") boardIds: List<Long>,
    ): List<BoardPostCount>

    @Query(
        """
        SELECT p.id
        FROM Post p
        WHERE p.clubMember.id IN :clubMemberIds
          AND p.isDeleted = false
          AND p.board.isDeleted = false
        ORDER BY p.id ASC
        """,
    )
    fun findActiveIdsByClubMemberIdIn(
        @Param("clubMemberIds") clubMemberIds: List<Long>,
    ): List<Long>

    @Query(
        """
        SELECT COUNT(p)
        FROM Post p
        WHERE p.clubMember.id IN :clubMemberIds
          AND p.isDeleted = false
          AND p.board.isDeleted = false
        """,
    )
    override fun countActiveByClubMemberIds(
        @Param("clubMemberIds") clubMemberIds: List<Long>,
    ): Long

    @EntityGraph(attributePaths = ["board", "board.club"])
    @Query(
        value = """
        SELECT p
        FROM Post p
        WHERE p.clubMember.user.id = :userId
          AND p.isDeleted = false
          AND p.board.isDeleted = false
        ORDER BY p.createdAt DESC, p.id DESC
        """,
        countQuery = """
        SELECT COUNT(p)
        FROM Post p
        WHERE p.clubMember.user.id = :userId
          AND p.isDeleted = false
          AND p.board.isDeleted = false
        """,
    )
    override fun findMyActivePosts(
        @Param("userId") userId: Long,
        pageable: Pageable,
    ): Page<Post>
}
